Improved priority inheritance implementation

sammanthp007 opened the pull request:
- [x] Implement the requirements of [this ticker](https://devel.rtems.org/ticket/2412) in RTEMS 4.11

gedare commented on cpukit/libmisc/cpuuse/cpuusagetop.c:
```diff
@@ -273,10 +273,10 @@ task_usage(Thread_Control* thread, void* arg)
             continue;
         case RTEMS_TOP_SORT_CURRENT_PRI:
-          if (thread->current_priority > data->tasks[j]->current_priority)
+          if (thread->Priority_node.current_priority > data->tasks[j]->Priority_node.current_priority)
```
Check for line lengths exceeding 80 characters, and wrap them if they do. This should be relatively easy to detect by skimming your commit files.

gedare replied:
https://devel.rtems.org/wiki/Developer/Coding/80_characters_per_line

gedare commented on cpukit/posix/src/pthread.c:
```diff
@@ -89,15 +89,16 @@ static bool _POSIX_Threads_Sporadic_budget_TSR_filter(
   return !_Thread_Owns_resources( the_thread ) &&
-    _Thread_Priority_less_than( the_thread->current_priority, *new_priority );
+    _Thread_Priority_less_than( the_thread->Priority_node.current_priority,
+            *new_priority );
```
See style guide for how to wrap, this is not correct, it should be more like
```C
    _Thread_Priority_less_than(
        the_thread->Priority_node.current_priority,
        *new_priority
    );

gedare replied:
Check for other places you have to fix line wrapping (lw)

gedare commented on cpukit/posix/src/pthread.c:
```diff
@@ -151,7 +152,8 @@ static bool _POSIX_Threads_Sporadic_budget_callout_filter(
   return !_Thread_Owns_resources( the_thread ) &&
-    _Thread_Priority_less_than( *new_priority, the_thread->current_priority );
+    _Thread_Priority_less_than( *new_priority,
+            the_thread->Priority_node.current_priority );
```
lw

gedare replied:
Put `);` on its own line.

gedare commented on cpukit/posix/src/pthreadgetschedparam.c:
```diff
@@ -51,7 +51,8 @@ int pthread_getschedparam(
         *param  = api->schedparam;
         param->sched_priority =
-          _POSIX_Priority_From_core( the_thread->current_priority );
+          _POSIX_Priority_From_core( the_thread->Priority_node.current_priority
```
lw

gedare commented on cpukit/rtems/src/tasksetpriority.c:
```diff
@@ -42,16 +42,12 @@ rtems_status_code rtems_task_set_priority(
       } else {
         *old_priority = _RTEMS_tasks_Priority_from_Core(
-          the_thread->current_priority
+         the_thread->Priority_node.current_priority
```
another indent for lw

gedare replied:
not enough indent level

gedare commented on cpukit/score/Makefile.am:
```diff
@@ -297,7 +297,8 @@ libscore_a_SOURCES += src/thread.c src/threadchangepriority.c \
     src/threadstackallocate.c src/threadstackfree.c src/threadstart.c \
-    src/threadstartmultitasking.c src/iterateoverthreads.c
+    src/threadstartmultitasking.c src/iterateoverthreads.c \
+	src/threadprioritynode.c
```
use spaces not tabs

gedare replied:
use space characters instead of tabs

gedare commented on cpukit/score/include/rtems/score/schedulersimpleimpl.h:
```diff
@@ -46,7 +46,7 @@ RTEMS_INLINE_ROUTINE bool _Scheduler_simple_Insert_priority_lifo_order(
   const Thread_Control *thread_next = (const Thread_Control *) next;
 
-  return thread_to_insert->current_priority <= thread_next->current_priority;
+  return thread_to_insert->Priority_node.current_priority <= thread_next->Priority_node.current_priority;
```
length

gedare commented on cpukit/score/src/coremutex.c:
```diff
@@ -58,7 +58,8 @@ CORE_mutex_Status _CORE_mutex_Initialize(
        */
-      if ( is_priority_ceiling && executing->current_priority < ceiling ) {
+      if ( is_priority_ceiling && executing->Priority_node.current_priority <
+              ceiling ) {
```
lw

gedare commented on cpukit/score/src/coremutex.c:
```diff
@@ -70,7 +71,7 @@ CORE_mutex_Status _CORE_mutex_Initialize(
        _Chain_Prepend_unprotected( &executing->lock_mutex,
                                    &the_mutex->queue.lock_queue );
-       the_mutex->queue.priority_before = executing->current_priority;
+       the_mutex->queue.priority_before = executing->Priority_node.current_priority;
```
length

gedare commented on cpukit/score/src/schedulerpriorityunblock.c:
```diff
@@ -51,10 +51,10 @@ Scheduler_Void_or_thread _Scheduler_priority_Unblock (
    *    a pseudo-ISR system task, we need to do a context switch.
    */
-  if ( the_thread->current_priority < _Thread_Heir->current_priority ) {
+  if ( the_thread->Priority_node.current_priority < _Thread_Heir->Priority_node.current_priority ) {
```
length

gedare replied:
line length

gedare commented on cpukit/score/src/thread.c:
```diff
@@ -29,8 +29,9 @@
-THREAD_OFFSET_ASSERT( real_priority );
+/* THREAD_OFFSET_ASSERT( Priority_node.current_priority );
+THREAD_OFFSET_ASSERT( Priority_node.real_priority );
+*/
```
Probably, this needs to be `THREAD_OFFSET_ASSERT( Priority_node );` because we replace the two fields of the struct by the one embedded struct.

gedare replied:
This should probably be `THREAD_OFFSET_ASSERT( Priority_node );`

gedare commented on cpukit/score/src/threadinitialize.c:
```diff
@@ -202,8 +202,11 @@ bool _Thread_Initialize(
+  _Chain_Set_off_chain( &the_thread->Priority_node.Node );
+  the_thread->Priority_node.current_priority        = priority;
+  the_thread->Priority_node.real_priority           = priority;
+  the_thread->Priority_node.waiting_to_hold  = NULL;
```
Align the = signs, at least for the ones you are adding.

gedare replied:
Align the = of the lines you add.

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -0,0 +1,184 @@
+/*
+ * Thread Priority Node
+ * 
+ * Copyright (C) 2017. Gedare Bloom and Samman Bikram Thapa.
```
Add a new copyright line for yourself, and make it also 2018, so...
`Copyright (C) 2018. Samman Bikram Thapa.`

gedare replied:
Add a new copyright line for yourself, and use the right year, e.g.,
` * Copyright (C) 2018. Samman Bikram Thapa.`

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -0,0 +1,184 @@
+    if ( current_priority < min_priority ) {
+        _Thread_Restore_priority(
+                the_thread
+                );
```
lw

gedare replied:
also, i'm not sure the line wrapping is needed here. the line shouldn't be too long I believe.

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -0,0 +1,184 @@
+        _Thread_Raise_priority(
+                the_thread,
+                min_priority
+                );
```
This one may be too long and line wrapping is needed, if so, align the closing `);` with the indentation level of the line starting this expression.

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -0,0 +1,184 @@
+  /* TODO: refactor? */
+  /* We are merely removing the priority node from 
+   * OwnerThread.Inherited_priorities
+   * and readding the the_thread->Priority_node back
```
typo: readding -> reading. also, can delete 'the'

gedare replied:
typo: readding -> reading. also can delete 'the'.

gedare commented on cpukit/libmisc/cpuuse/cpuusagetop.c:
```diff
@@ -273,10 +273,10 @@ task_usage(Thread_Control* thread, void* arg)
             continue;
         case RTEMS_TOP_SORT_CURRENT_PRI:
-          if (thread->current_priority > data->tasks[j]->current_priority)
+          if (thread->Priority_node.current_priority > data->tasks[j]->Priority_node.current_priority)
```
Lines too long. Fix everywhere. It should be easy to find if you look at the commit files.
https://devel.rtems.org/wiki/Developer/Coding/80_characters_per_line

gedare commented on cpukit/posix/src/pthread.c:
```diff
@@ -89,15 +89,16 @@ static bool _POSIX_Threads_Sporadic_budget_TSR_filter(
   return !_Thread_Owns_resources( the_thread ) &&
-    _Thread_Priority_less_than( the_thread->current_priority, *new_priority );
+    _Thread_Priority_less_than( the_thread->Priority_node.current_priority,
+            *new_priority );
```
Wrong line wrapping. There should not be anything after the opening parens on the first line, and the closing parens should be aligned with the first line. e.g.,:
```C
    _Thread_Priority_less_than(
        the_thread->Priority_node.current_priority,
        *new_priority
    );
```
Fix this everywhere.

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -0,0 +1,184 @@
+    if ( current_priority < min_priority ) {
+        _Thread_Restore_priority(
+                the_thread
+                );
```
broken line wrapping

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -0,0 +1,184 @@
+  mutex = _Thread_Dequeue_priority_node( &the_thread->Priority_node );
+  if ( mutex != NULL ) {
+    mutex_owner_needs_update = _Thread_Enqueue_priority_node( the_thread, mutex
+            );
```
line wrapping

sammanthp007 replied:
these were due to my vim config.. fixing it now.

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -0,0 +1,184 @@
+  if ( mutex != NULL ) {
+    mutex_owner_needs_update = _Thread_Enqueue_priority_node( the_thread, mutex
+            );
+    /* Samman TODO: Should not we update mutex.current_priority? */
```
Delete TODO. There is no mutex.current_priority field.

sammanthp007 replied:
Oops.. that should not be there.

gedare commented on testsuites/sptests/spsem05/init.c:
```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2017 Gedare Bloom, Samman Bikram Thapa.
```
fix

gedare commented:
Also, you should probably apply my patches/changes with my authorship info, and your code changes with your info.

gedare commented on cpukit/score/src/threadsetpriority.c:
```diff
@@ -30,13 +30,15 @@ static bool _Thread_Set_priority_filter(
 
-  old_priority_ptr = arg;
-  *old_priority_ptr = current_priority;
+  if (arg != NULL) {
```
Is this NULL check required?

sammanthp007 replied:
Nope. You are correct. The check is not required.

sammanthp007 commented:
Dr. @gedare Bloom, 
- [x] updated the style issues
- [x] rebased to have your code be associated with your authorship info
- [x] removed typos from git commit messages

gedare commented on cpukit/libmisc/cpuuse/cpuusagetop.c:
```diff
@@ -273,10 +273,12 @@ task_usage(Thread_Control* thread, void* arg)
         case RTEMS_TOP_SORT_REAL_PRI:
-          if (thread->Priority_node.real_priority > data->tasks[j]->Priority_node.real_priority)
+          if (thread->Priority_node.real_priority >
+                  data->tasks[j]->Priority_node.real_priority)
```
i don't know if you got my other messages, but this should be
```C
          if (
                  thread->Priority_node.real_priority >
                  data->tasks[j]->Priority_node.real_priority
          )
```

gedare commented on cpukit/posix/src/pthread.c:
```diff
@@ -97,8 +97,9 @@ static bool _POSIX_Threads_Sporadic_budget_TSR_filter(
-            *new_priority );
+    _Thread_Priority_less_than(
+        the_thread->Priority_node.current_priority,
+        *new_priority );
```
put `);` on own line, aligned with first line
```C
    _Thread_Priority_less_than(
        the_thread->Priority_node.current_priority,
        *new_priority
    );
```

gedare commented on cpukit/posix/src/pthreadgetschedparam.c:
```diff
@@ -51,8 +51,9 @@ int pthread_getschedparam(
-                  );
+          _POSIX_Priority_From_core(
+                the_thread->Priority_node.current_priority
+                );
```
Align `);` with _POSIX_

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -1,7 +1,7 @@
  * Thread Priority Node
  * 
- * Copyright (C) 2017. Gedare Bloom and Samman Bikram Thapa.
+ * Copyright (C) 2018. Samman Bikram Thapa.
```
Don't delete my copyright.

gedare commented on testsuites/sptests/spsem05/init.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017 Gedare Bloom, Samman Bikram Thapa.
+ * Copyright (c) 2018 Gedare Bloom, Samman Bikram Thapa.
```
Split the copyright lines.

gedare commented on cpukit/score/src/schedulersimpleunblock.c:
```diff
@@ -43,10 +43,10 @@ Scheduler_Void_or_thread _Scheduler_simple_Unblock(
    *    a pseudo-ISR system task, we need to do a context switch.
    */
-  if ( the_thread->current_priority < _Thread_Heir->current_priority ) {
+  if ( the_thread->Priority_node.current_priority < _Thread_Heir->Priority_node.current_priority ) {
```
line length

gedare commented on cpukit/libmisc/cpuuse/cpuusagetop.c:
```diff
@@ -273,10 +273,16 @@ task_usage(Thread_Control* thread, void* arg)
         case RTEMS_TOP_SORT_REAL_PRI:
-          if (thread->Priority_node.real_priority > data->tasks[j]->Priority_node.real_priority)
+          if (
+                thread->Priority_node.real_priority >
```
This is better,but I think you indented it too far here, and a few other places. It should be 4 white spaces deeper than the if line, e.g.:
```C
if (
    thread->Priority_node.real_priority >
    data->tasks[j]->Priority_node.real_priority
)

gedare commented on testsuites/sptests/sp34/changepri.c:
```diff
@@ -43,8 +43,11 @@ const char *CallerName(void)
+  sprintf( buffer,
+          "0x%08x -- %d",
+          rtems_task_self(),
+          executing->Priority_node.current_priority
```
Here, put buffer on a separate line, and align with 4 indents, like:
```C
sprintf(
    buffer,
    "0x%08x -- %d",
    rtems_task_self(),
    executing->Priority_node.current_priority
);
```

gedare commented on testsuites/sptests/spsem05/init.c:
```diff
@@ -1,5 +1,6 @@
 /*
- * Copyright (c) 2017 Gedare Bloom, Samman Bikram Thapa.
+ * Copyright (C) 2018. Gedare Bloom.
```
Leave the year of my copyright as 2017.

gedare commented on cpukit/score/src/threadprioritynode.c:
```diff
@@ -1,7 +1,8 @@
  * Thread Priority Node
  * 
- * Copyright (C) 2017. Gedare Bloom and Samman Bikram Thapa.
+ * Copyright (C) 2018. Gedare Bloom.
```
Leave the year of my copyright as 2017.

gedare commented on cpukit/score/src/coremutex.c:
```diff
@@ -58,8 +58,10 @@ CORE_mutex_Status _CORE_mutex_Initialize(
-              ceiling ) {
+      if (
+            is_priority_ceiling &&
+            executing->Priority_node.current_priority < ceiling
```
check indent level

gedare commented on cpukit/posix/src/pthreadgetschedparam.c:
```diff
@@ -51,8 +51,9 @@ int pthread_getschedparam(
-          _POSIX_Priority_From_core( the_thread->Priority_node.current_priority
-                  );
+          _POSIX_Priority_From_core(
+                the_thread->Priority_node.current_priority
```
check indent level

gedare commented on cpukit/libmisc/cpuuse/cpuusagetop.c:
```diff
@@ -273,10 +273,16 @@ task_usage(Thread_Control* thread, void* arg)
-          if (thread->Priority_node.current_priority > data->tasks[j]->Priority_node.current_priority)
+          if (
+                thread->Priority_node.current_priority >
+                data->tasks[j]->Priority_node.current_priority
```
ditto.

gedare commented:
I manually merged this with a few small changes to your clean-up commit.
